### DISCUSSION TITLE
Remote Monitoring behind Dynamic DNS

### DISCUSSION THREAD

**Discussion by NeilBain** [wills106/homeassistant-solax-modbus#176] [Q&A]
First off... a huge thanks for this integration... it is VERY impressive. 

My home-assistant server is running on Ubuntu on a free Oracle cloud instance which communicates with my modbus module via my public IP address with the appropriate address translation.  This works really well for me except for one niggle.  The home assistant-solax-modbus setup assumes an IP address (xxx.xxx.xxx.xxx) and does not accept a domain name (example.com).  Although I have a dynamic DNS service available, I cannot use it for this purpose.

My local IP address changes regularly enough for this to be a minor hassle to reconfigure the integration each time.

Is there any way to get the integration to accept a domain name?

If not, (and I accept this would now be off topic) any thought on how I might redirect traffic such that I can use the dynamic DNS service.

With thanks

**Comment by wills106** [comment]
It's not something I have ever considered support to be honest. When filling out the address it's set to reject any characters.
I was expecting this to used locally only as there is no security to Modbus, so if you expose your Waveshare or other device to the Internet you run the risk of other people being able to control your Inverter.

**Reply by NeilBain** [reply]
Thanks @wills106 

Yes, security issues well understood with mitigations built into my firewall. (I am more worried about the control that can be applied via the Solax Pocket WiFi / SolaxCloud... but that's a different conversation!)

I suspect I can solve this outside of the integration via outbound address translation firewall rules....  but thanks for the response.

**Reply by wills106** [reply]
You could manually remove the "disallowed" section from the config flow? But you would have to perform that each time you upgrade the Integration.

Instead of exposing the Waveshare type device to the Internet, could you setup a VPN? Or doesn't your router / firewall  support that?

**Reply by infradom** [reply]
If I remember well, I once tried to allow dns names, by modifiying the config flow. For some reason I cannot remember, I did not succeed and gave up quickly. Maybe I should have tried harder. 
Anyway, it might bemore complicated, as each time the dns to ip mapping changes, the connection to the inverter needs to be re-created (maybe this is already the case, but I would not bet on it).
If I find some time, I might try again ...
Anyway as @wills106 explained, we have to be aware of the security risks if we use the Internet.

**Reply by NeilBain** [reply]
Giving this more thought, I now feel strongly I should not implement this within the integration (by modifying the config flow for example) as it would cause a DNS look-up on every modbus request.  For me, this would be every 7 seconds and could represent a significant performance bottle neck (a delayed DNS response would delay the whole modbus call).

Instead, I am going to:
- implement an IPTables rule to NAT to my IP address
- do periodic DNS lookups from the OS (`dig +short my_ddns.com`)
- compare this to previous lookups
- if changed, modify the saved IPtables rules (using sed)
- reload the table

Thank you everyone for helping my thinking :-)
